rust/src/lib.rs

```diff
@@ -180,34 +212,42 @@ Accept-Encoding: gzip,deflate\r
             let (header, value) = split2(&line, ":").ok_or(MALFORMED_RESPONSE)?;
             let header = header.to_ascii_lowercase();
             let value = value.trim();
-            if header == "content-encoding" {
-                encoding = value.parse().map_err(|_| UNSUPPORTED_ENCODING)?;
-            }
             headers.insert(header, value.to_string());
         }
 
-        // 11. Read body
-        let mut body = Vec::new();
-        match encoding {
-            ContentEncoding::Gzip => {
-                GzDecoder::new(reader)
-                    .read_to_end(&mut body)
-                    .map_err(|_| MALFORMED_RESPONSE)?;
-            }
-            ContentEncoding::Deflate => {
-                DeflateDecoder::new(reader)
-                    .read_to_end(&mut body)
-                    .map_err(|_| MALFORMED_RESPONSE)?;
-            }
-            ContentEncoding::Identity => {
-                reader
-                    .read_to_end(&mut body)
-                    .map_err(|_| MALFORMED_RESPONSE)?;
-            }
-            _ => {
-                panic!("{}", UNSUPPORTED_ENCODING);
+        let content_encoding: ContentEncoding = match headers.get("content-encoding") {
+            Some(encoding) => encoding.parse().map_err(|_| UNSUPPORTED_ENCODING)?,
+            None => ContentEncoding::Identity,
+        };
+
+        let body = match headers.get("transfer-encoding") {
+            Some(encoding) => {
+                let mut unchunked = Vec::new();
+                if "chunked".eq_ignore_ascii_case(encoding) {
+                    loop {
+                        let mut line = String::new();
+                        reader
+                            .read_line(&mut line)
+                            .map_err(|_| MALFORMED_RESPONSE)?;
+                        let n_bytes = i64::from_str_radix(line.trim_end(), 16).unwrap_or(0);
+                        if n_bytes == 0 {
+                            break;
+                        }
+                        let mut chunk = vec![0u8; n_bytes as usize];
+                        reader
+                            .read_exact(&mut chunk)
+                            .map_err(|_| MALFORMED_RESPONSE)?;
+                        reader.read_exact(&mut vec![0u8; 2]).unwrap();
+                        unchunked.write_all(&chunk).unwrap();
+                    }
```

Comment:
if you want, writing `ChunkReader` with `impl Read for ChunkReader` will provide a nice way to make `Read` to `Read` conversion.

Reply:
I will try